File: gosi-kafka-sdk-dotnet/Auth/AuthErrorClassifier.cs
namespace Gosi.Kafka.Sdk.Auth;

using System;
using Confluent.Kafka;

public enum AuthErrorType
{
    AUTHENTICATION_FAILURE,
    AUTHORIZATION_DENIED,
    UNKNOWN_ERROR
}

public static class AuthErrorClassifier
{
    public static AuthErrorType Classify(Exception exception)
    {
        if (exception == null) return AuthErrorType.UNKNOWN_ERROR;

        // Confluent.Kafka surfaces auth errors via ConsumeException/ProduceException which wrap KafkaException
        // We look at the Error.Code to distinguish Authentication vs Authorization

        if (exception is KafkaException kafkaException)
        {
            if (kafkaException.Error.Code == ErrorCode.Local_Authentication ||
                kafkaException.Error.Code == ErrorCode.SaslAuthenticationFailed)
            {
                return AuthErrorType.AUTHENTICATION_FAILURE;
            }

            if (kafkaException.Error.Code == ErrorCode.TopicAuthorizationFailed ||
                kafkaException.Error.Code == ErrorCode.GroupAuthorizationFailed ||
                kafkaException.Error.Code == ErrorCode.ClusterAuthorizationFailed)
            {
                return AuthErrorType.AUTHORIZATION_DENIED;
            }
        }
        
        // Unwrap AggregateException if present
        if (exception is AggregateException aggregateException && aggregateException.InnerException != null)
        {
            return Classify(aggregateException.InnerException);
        }

        // Specifically check for InnerExceptions that might be KafkaException
        if (exception.InnerException != null)
        {
            return Classify(exception.InnerException);
        }

        return AuthErrorType.UNKNOWN_ERROR;
    }

    public static Exception ClassifyAndWrap(Exception exception)
    {
        var type = Classify(exception);
        if (type == AuthErrorType.UNKNOWN_ERROR)
        {
            return exception;
        }

        return new GosiAuthException(type, exception.Message, exception);
    }
}

public class GosiAuthException : Exception
{
    public AuthErrorType ErrorType { get; }

    public GosiAuthException(AuthErrorType errorType, string message, Exception innerException) 
        : base($"[{errorType}] {message}", innerException)
    {
        ErrorType = errorType;
    }
}

File: gosi-kafka-sdk-java/gosi-kafka-sdk-core/src/main/java/com/gosi/kafka/sdk/auth/AuthErrorClassifier.java
package com.gosi.kafka.sdk.auth;

import org.apache.kafka.common.errors.ClusterAuthorizationException;
import org.apache.kafka.common.errors.GroupAuthorizationException;
import org.apache.kafka.common.errors.SaslAuthenticationException;
import org.apache.kafka.common.errors.TopicAuthorizationException;
import org.apache.kafka.common.errors.TransactionalIdAuthorizationException;
import org.apache.kafka.common.errors.DelegationTokenAuthorizationException;

/**
 * Classifies Kafka exceptions into distinct auth error types.
 * <p>
 * The Enterprise Event Streaming Framework requires that the SDK surface
 * authorization denials distinctly from authentication failures. Broker-side
 * authorization is two-tier (Kafka ACLs first, RBAC via Metadata Service on a miss)
 * — the SDK does not implement this, but must report the failure class correctly.
 * </p>
 */
public final class AuthErrorClassifier {

    private AuthErrorClassifier() {
        // Utility class
    }

    /**
     * Classifies a Throwable into an {@link AuthErrorType}.
     *
     * @param throwable the exception to classify
     * @return the classified error type
     */
    public static AuthErrorType classify(Throwable throwable) {
        if (throwable == null) {
            return AuthErrorType.UNKNOWN;
        }

        // Direct match
        AuthErrorType type = classifyDirect(throwable);
        if (type != AuthErrorType.UNKNOWN) {
            return type;
        }

        // Check cause chain
        Throwable cause = throwable.getCause();
        int depth = 0;
        while (cause != null && depth < 10) {
            type = classifyDirect(cause);
            if (type != AuthErrorType.UNKNOWN) {
                return type;
            }
            cause = cause.getCause();
            depth++;
        }

        return AuthErrorType.UNKNOWN;
    }

    /**
     * Wraps the given throwable as a {@link GosiAuthException} with the classified type.
     *
     * @param throwable the original exception
     * @return a GosiAuthException wrapping the original error
     */
    public static GosiAuthException classifyAndWrap(Throwable throwable) {
        AuthErrorType type = classify(throwable);
        String message = buildMessage(type, throwable);
        return new GosiAuthException(type, message, throwable);
    }

    /**
     * Returns true if the given throwable represents a security-related error
     * (either authentication or authorization).
     */
    public static boolean isSecurityError(Throwable throwable) {
        AuthErrorType type = classify(throwable);
        return type == AuthErrorType.AUTHENTICATION_FAILURE || type == AuthErrorType.AUTHORIZATION_DENIED;
    }

    private static AuthErrorType classifyDirect(Throwable throwable) {
        // Authentication failures — SASL handshake, JWT validation
        if (throwable instanceof SaslAuthenticationException) {
            return AuthErrorType.AUTHENTICATION_FAILURE;
        }

        // Authorization denials — ACL/RBAC evaluated by broker
        if (throwable instanceof TopicAuthorizationException) {
            return AuthErrorType.AUTHORIZATION_DENIED;
        }
        if (throwable instanceof GroupAuthorizationException) {
            return AuthErrorType.AUTHORIZATION_DENIED;
        }
        if (throwable instanceof ClusterAuthorizationException) {
            return AuthErrorType.AUTHORIZATION_DENIED;
        }
        if (throwable instanceof TransactionalIdAuthorizationException) {
            return AuthErrorType.AUTHORIZATION_DENIED;
        }
        if (throwable instanceof DelegationTokenAuthorizationException) {
            return AuthErrorType.AUTHORIZATION_DENIED;
        }

        // Heuristic: check error message for auth-related keywords
        String message = throwable.getMessage();
        if (message != null) {
            String lowerMessage = message.toLowerCase();
            if (lowerMessage.contains("authentication failed") || lowerMessage.contains("sasl authentication")) {
                return AuthErrorType.AUTHENTICATION_FAILURE;
            }
            if (lowerMessage.contains("not authorized") || lowerMessage.contains("authorization failed")) {
                return AuthErrorType.AUTHORIZATION_DENIED;
            }
        }

        return AuthErrorType.UNKNOWN;
    }

    private static String buildMessage(AuthErrorType type, Throwable throwable) {
        String originalMessage = throwable != null ? throwable.getMessage() : "unknown";
        switch (type) {
            case AUTHENTICATION_FAILURE:
                return "Authentication failure: " + originalMessage;
            case AUTHORIZATION_DENIED:
                return "Authorization denied (ACL/RBAC): " + originalMessage;
            case TRANSIENT_ERROR:
                return "Transient error: " + originalMessage;
            default:
                return "Unclassified error: " + originalMessage;
        }
    }
}
